### PR TITLE
sunlineSRuKF module

### LINKED ISSUE #217: sunlineSRuKF
**Describe your use case**
Implementing a square root unscented kalman filter that inherits from the square root UKF interface to estimate the sun heading using coarse sun sensor measurements.

**Describe alternatives solutions you've considered**
A simple weighted least squares approach to estimate the sun heading fails in the case of limited sky coverage, because whenever the sun is lost during maneuvering the LS algorithm does not keep memory of the sun heading.

### PR DESCRIPTION
**Pull request by rcalaon**
* **Tickets addressed:** bsk-793
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description
Square root unscented kalman filter to estimate the sunline direction in body-frame coordinates using a combination of coarse sun sensor and rate gyro measurements.

## Verification
The unit test tests the correctness of the propagation, the measurement processing, and state output of the filter compared to the measurement noise of the sensors.

## Documentation
Documentation is provided.

## Future work
The input navigation message will be substituted with input gyro message.

### CODE REVIEW

**Review by thibaudteil — APPROVED** [review]
I think Pat likes to squash all of the commits into one single huge one when PR'ing a new module, but I'm ok either way

File: src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.cpp [hunk 1]
```diff
@@ -67,3 +67,50 @@ Eigen::VectorXd SunlineSRuKF::propagate(std::array<double, 2> interval, const Ei
 
     return X;
 }
+
+/*! Set the CSS measurement noise
+    @param double measurementNoiseScale
```
Comment (patkenneally) [inline src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.cpp]: :large_orange_diamond: Looks like this comment is a copy and paste for the next three functions. Can change to match the function param if you like.

File: src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.cpp [hunk 2]
```diff
@@ -67,3 +67,50 @@ Eigen::VectorXd SunlineSRuKF::propagate(std::array<double, 2> interval, const Ei
...
+    @return void
+    */
+void SunlineSRuKF::setCssMeasurementNoiseStd(const double cssMeasurementNoise) {
```
Comment (patkenneally) [inline src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.cpp]: :large_orange_diamond: Is this setting the standard deviation?

File: src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.h [hunk 3]
```diff
@@ -54,6 +54,8 @@ class SunlineSRuKF: public SRukfInterface {
 
     double sensorUseThresh = 0;  //!< Threshold below which we discount sensors
     double measNoiseScaling = 1; //!< [s] Scale factor that can be applied on the measurement noise to over/under weight
+    double cssMeasNoise = 0;     //!< [-] CSS measurement noise std
```
Comment (patkenneally) [inline src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.h]: :large_orange_diamond: If this is a standard deviation should we indicate this in variable name?

File: src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.cpp [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+    auto gyroMeasurements = Measurement();
+    gyroMeasurements.validity = true;
+    gyroMeasurements.name = "gyro";
+    gyroMeasurements.size = 3;
+    gyroMeasurements.timeTag = navAttInputBuffer.timeTag;
+    gyroMeasurements.observation = cArray2EigenVector3d(navAttInputBuffer.omega_BN_B);
+    gyroMeasurements.model = lastThreeStates;
+    gyroMeasurements.noise.resize(3, 3);
+    Eigen::MatrixXd I(3,3);
```
Comment (patkenneally) [inline src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.cpp]: :large_orange_diamond: Could this become `Eigen::Matrix3d I = Eigen::Matrix3d::Identity();`. This means that a variable is declared and initialized on the same line.

File: src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.h [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+ MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+ */
+
+
+/*! @brief Top level structure for the flyby OD unscented kalman filter.
```
Comment (patkenneally) [inline src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.h]: :small_red_triangle_down: Delete this comment

File: src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.cpp [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+SunlineSRuKF::SunlineSRuKF() = default;
+
+SunlineSRuKF::~SunlineSRuKF() = default;
+
+/*! Initialize C-wrapped output messages */
+void SunlineSRuKF::SelfInit(){
+    NavAttMsg_C_init(&this->navAttOutMsgC);
+}
+
+/*! Reset the flyby OD filter to an initial state and
```
Comment (patkenneally) [inline src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.cpp]: :small_red_triangle_down: Check all comments and function headers are accurate and descriptive.

**Review by patkenneally — APPROVED** [review]
:small_red_triangle_down: *Issues* to address before merge
:large_orange_diamond: *Requests* that should not block merge, but should at least be discussed
:large_blue_circle: *Recommendations* that can be ignored if desired

File: src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.h (hunk 5 above)
Comment (rcalaon) [reply src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.h]: Done

File: src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.cpp (hunk 6 above)
Comment (rcalaon) [reply src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.cpp]: Done

File: src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.cpp (hunk 4 above)
Comment (rcalaon) [reply src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.cpp]: Done

File: src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.h (hunk 3 above)
Comment (rcalaon) [reply src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.h]: I've added the Std to all the related variable names

File: src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.cpp (hunk 2 above)
Comment (rcalaon) [reply src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.cpp]: Yes

File: src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.cpp (hunk 1 above)
Comment (rcalaon) [reply src/fswAlgorithms/attDetermination/sunlineSRuKF/sunlineSRuKF.cpp]: Fixed it

**Pull request merged**